hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeApplications    #-}
+
+{-|
+Description : Allowed effects for interacting with Nar files.
+Maintainer  : Shea Levy <shea@shealevy.com>
+|-}
+module System.Nix.Nar (
+    FileSystemObject(..)
+  , IsExecutable (..)
+  , Nar(..)
+  , getNar
+  , localUnpackNar
+  , narEffectsIO
+  , putNar
+  ) where
+
+import           Control.Applicative
+import           Control.Monad              (replicateM, replicateM_)
+import Control.Monad.Trans
+import           Control.Monad.Writer
+import qualified Data.Binary                as B
+import qualified Data.Binary.Get            as B
+import qualified Data.Binary.Put            as B
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import           Data.Foldable              (forM_)
+import qualified Data.Map                   as Map
+import           Data.Maybe                 (fromMaybe)
+import           Data.Monoid                ((<>))
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           GHC.Int                    (Int64)
+import           System.Directory
+import           System.FilePath
+import           System.Posix.Files         (createSymbolicLink)
+
+import           System.Nix.Path
+
+
+data NarEffects (m :: * -> *) = NarEffects {
```

Comment:
Maybe add a TODO for separating out read effects from write effects. Certainly not necessary to start though.